File: Best time to buy and sell stock.java
// Leetcode Problem 121: Best Time to Buy and Sell Stock
// JAVA CODE
class Solution {
    public int maxProfit(int[] prices) {
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;

        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price; // Update the lowest buying price
            } else if (price - minPrice > maxProfit) {
                maxProfit = price - minPrice; // Update maximum profit
            }
        }

        return maxProfit;
    }
}

File: Best time to buy and sell stock.py
# Leetcode Problem 121: Best Time to Buy and Sell Stock
# PYTHON CODE
class Solution:

    def maxProfit(self, prices: list[int]) -> int:
        min_price = float("inf")
        max_profit = 0

        for price in prices:
            if price < min_price:
                min_price = price  # Update the lowest buy price
            elif price - min_price > max_profit:
                max_profit = price - min_price  # Update maximum profit

        return max_profit